src/extension.ts

```diff
@@ -60,6 +61,13 @@ export async function activate(context: vscode.ExtensionContext): Promise<any> {
 		}
 	}
 
+	if (config('terraform').has('languageServer.version')) {
+		const langServerVer = config('terraform').get('languageServer.version', defaultVersionString)
+		if (!isValidVersionString(langServerVer)) {
+			vscode.window.showWarningMessage(`The Terraform Language Server Version string '${langServerVer}' is not a valid semantic version and will be ignored.`);
+		}
+	}
```

Comment:
Thank you for adding the validation!

Is there a particular reason why do it here, instead of prior to updating below? 🤔 

Relatedly I think that the user would benefit from even more validation - e.g.

1. check for conflict between `pathToBinary` and the version
2. check that there is at least 1 version which satisfies the constraint

Re (2) the current UX when I try to set a constraint which doesn't satisfy any version is not ideal:

![Screenshot 2021-06-17 at 16 33 50](https://user-images.githubusercontent.com/287584/122428482-d53dd300-cf89-11eb-9eef-89aac1a75368.png)

Reply:
> Is there a particular reason why do it here, instead of prior to updating below? 

There feels like there are two layers of validation here..

**"Static" Validation** (Doesn't need to call out to "anything")
Is the `required_version` string even a valid semver string?
Did the user also specify a conflicting setting e.g. `pathToBinary`?

**"Dynamic" Validation** (Needs to call out to "something" for validation)
Does the valid semver string constraint actually return at least 1 possible version.?

So the reason I "static" validation at the top was to validate the settings as soon as possible and prompt the user. Note that this validation is synchronous and blocking the extension activation so it needs to be fast. (TODO - Need to check if `showWarningMessage` blocks the extension activation.  Don't think so)

Now, as you suggested, I could move the validation to be later...that's fine. This is just the approach I took.

Reply:
> Re (2) the current UX when I try to set a constraint which doesn't satisfy any version is not ideal:

UX wise we are constrained by VScode. 

Interesting... that UI isn't from the updater code....

Ahh ok... it happens when Fresh install of extension and a valid semver string but has no versions which satisfies the constraint.

Yeah, may need to change the code  to request the default version if  no applicable version could be found AND there is no existing install of the LS.

Reply:
Ok... so now it will attempt to find the required_version, and fall back to the default version, and then disable updating if it can't find any version.

This doesn't fix another problem I _think_ I see, but that was already there.

* First time user with the extension installed (but not activated)
* ... with no internet connection (or blocked getting to the API endpoint)
* ... It can't download the LSP, so the extension will throw the error.

Reply:
Yes - I am aware this is a problem - or that the UX is confusing at the least - I already started working on one part of the problem in https://github.com/hashicorp/vscode-terraform/pull/691

I reckon there's still more to address in that area though.

Reply:
I noticed that you've now got a Client Handler, so in theory "we/you" should be able to start the extension and NOT the language server/client part so you can degrade gracefully.  Definitely on the right track!